File: frontend/src/types/structure.ts
export type ContentBlockType = 'text' | 'code' | 'table' | 'image';

export interface BoundingBox {
  left: number;
  top: number;
  width: number;
  height: number;
}

export interface ContentBlock {
  id: string;
  type: ContentBlockType;
  content: string;
  orderIndex: number;
  wordCount?: number;
  bbox?: BoundingBox;
}

export interface Section {
  id: string;
  heading: string;
  level: number; // 2 = Heading 2, 3 = Heading 3
  contentBlocks: ContentBlock[];
  orderIndex: number;
  wordCount?: number;
}

export interface Chapter {
  chapter_id: string;
  heading: string;
  content_blocks: ContentBlock[];
  sections?: Section[];
  order_index: number;
  wordCount?: number;
}

export interface DocumentStructure {
  documentId: string;
  source: 'docx' | 'pdf' | 'textract';
  chapters: Chapter[];
  totalWordCount?: number;
  pageCount?: number;
}

File: backend/app/schemas/structure.py
from pydantic import BaseModel
from typing import Literal


class BoundingBox(BaseModel):
    """Optional geometry for chip/raw-text layout (e.g. from OCR or PyMuPDF)."""
    left: float = 0
    top: float = 0
    width: float = 0
    height: float = 0


class ContentBlock(BaseModel):
    id: str
    type: Literal["text", "code", "table", "image"]
    content: str
    orderIndex: int
    wordCount: int | None = None
    bbox: BoundingBox | None = None  # optional geometry for raw-text chip view


class Section(BaseModel):
    id: str
    heading: str
    level: int
    contentBlocks: list[ContentBlock]
    orderIndex: int
    wordCount: int | None = None


class Chapter(BaseModel):
    chapter_id: str
    heading: str
    content_blocks: list[ContentBlock]
    sections: list[Section] | None = None
    order_index: int
    wordCount: int | None = None


class DocumentStructure(BaseModel):
    documentId: str
    source: Literal["docx", "pdf", "textract", "ocr"]
    chapters: list[Chapter]
    totalWordCount: int | None = None
    pageCount: int | None = None
